modules/images/image-loading-optimization/class-ilo-url-metrics-group.php

```diff
@@ -0,0 +1,172 @@
+ * @access private
+ */
+final class ILO_URL_Metrics_Group {
+
+	/**
+	 * URL metrics.
+	 *
+	 * @var ILO_URL_Metric[]
+	 */
+	private $url_metrics;
+
+	/**
+	 * Minimum possible viewport width for the group (inclusive).
+	 *
+	 * @var int
+	 * @phpstan-var int<0, max>
+	 */
+	private $minimum_viewport_width;
+
+	/**
+	 * Maximum possible viewport width for the group (inclusive).
+	 *
+	 * @var int
+	 * @phpstan-var int<1, max>
+	 */
+	private $maximum_viewport_width;
+
+	/**
+	 * Sample size for URL metrics for a given breakpoint.
+	 *
+	 * @var int
+	 * @phpstan-var int<1, max>
+	 */
+	private $sample_size;
+
+	/**
+	 * Freshness age (TTL) for a given URL metric.
+	 *
+	 * @var int
+	 * @phpstan-var int<0, max>
+	 */
+	private $freshness_ttl;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param ILO_URL_Metric[] $url_metrics            URL metrics to add to the group.
+	 * @param int              $minimum_viewport_width Minimum possible viewport width for the group.
+	 * @param int              $maximum_viewport_width Maximum possible viewport width for the group.
+	 * @param int              $sample_size            Sample size for the maximum number of viewports in a group between breakpoints.
+	 * @param int              $freshness_ttl          Freshness age (TTL) for a given URL metric.
+	 */
+	public function __construct( array $url_metrics, int $minimum_viewport_width, int $maximum_viewport_width, int $sample_size, int $freshness_ttl ) {
+		$this->url_metrics            = $url_metrics;
+		$this->minimum_viewport_width = $minimum_viewport_width;
+		$this->maximum_viewport_width = $maximum_viewport_width;
+		$this->sample_size            = $sample_size;
+		$this->freshness_ttl          = $freshness_ttl;
+	}
+
+	/**
+	 * Gets the minimum possible viewport width (inclusive).
+	 *
+	 * @return int Minimum viewport width.
+	 */
+	public function get_minimum_viewport_width(): int {
+		return $this->minimum_viewport_width;
+	}
+
+	/**
+	 * Gets the maximum possible viewport width (inclusive).
+	 *
+	 * @return int Minimum viewport width.
+	 */
+	public function get_maximum_viewport_width(): int {
+		return $this->maximum_viewport_width;
+	}
+
+	/**
+	 * Checks whether the provided viewport width is within the minimum/maximum range for
+	 *
+	 * @param int $viewport_width Viewport width.
+	 * @return bool Whether the viewport width is in range.
+	 */
+	public function is_viewport_width_in_range( int $viewport_width ): bool {
+		return (
+			$viewport_width >= $this->minimum_viewport_width &&
+			$viewport_width <= $this->maximum_viewport_width
+		);
+	}
+
+	/**
+	 * Adds a URL metric to the group if it is in the bounds of the max and min viewport width.
+	 *
+	 * @param ILO_URL_Metric $url_metric URL metric.
+	 * @return bool Whether the URL metric was added.
+	 */
+	public function add_url_metric( ILO_URL_Metric $url_metric ): bool {
+		if ( ! $this->is_viewport_width_in_range( $url_metric->get_viewport()['width'] ) ) {
+			return false;
+		}
```

Comment:
Rather than checking this here and returning a boolean, a more explicit API would be to check whether `is_viewport_width_in_range` in the `URL_Metrics_Group_Collection` class _before_ calling `add_url_metric`, with the method here throwing an exception if the given metric doesn't fit in the group.

This prevents mistakes in a way the developer will for sure find out about it. Returning a success/failure value from a method that does something other than check for a specific thing isn't very intuitive and often may go ignored. Think about all the `register_post_type()` calls in plugins and how uncommon it is to check for the return value - even though there is one.

Reply:
OK, I've implemented this in c362a6f42f7f9cb92bc969d93ff952dbf17c0c67. Nevertheless, a URL metric should always fit into a group contained in a collection because when the collection creates the groups it will ensure the groups span all non-negative integers.

Reply:
Right. This is just to make a clean API, should someone try to do something custom with it. Plus for any potential future iterations within the plugin code itself.